src/common/src/array/stream_chunk.rs

```diff
@@ -330,17 +330,32 @@ impl StreamChunk {
             if !c.vis(curr) {
                 continue;
             }
-            if let Some(prev) = prev_r {
-                if matches!(c.op(prev), Op::UpdateDelete | Op::Delete)
-                    && matches!(c.op(curr), Op::UpdateInsert | Op::Insert)
-                    && c.row_ref(prev) == c.row_ref(curr)
-                {
-                    c.set_vis(prev, false);
-                    c.set_vis(curr, false);
-                    prev_r = None;
-                } else {
-                    prev_r = Some(curr)
-                }
+            if let Some(prev) = prev_r
+                && (
+                    // 1. Delete then Insert
+                    (matches!(c.op(prev), Op::UpdateDelete | Op::Delete)
+                    && matches!(c.op(curr), Op::UpdateInsert | Op::Insert))
+                    ||
+                    // 2. Insert then Delete
```

Comment:
This makes sense to me. I wonder why we don't handle this elimination previously.

Reply:
I guess by "updating", the first impression comes to our mind is `UpdateDelete` then `UpdateInsert`.

However, there are indeed also cases like https://github.com/risingwavelabs/risingwave/issues/17450 where we first issue an `Insert` of an all-NULL _unmatched_ row in outer join, then immediately retract it. Currently it's handled by another elimination implementation, but it seems that we can make it more general purpose.